Libraries/D3D12RaytracingFallback/src/TreeletReorder.cpp

```diff
@@ -55,27 +59,54 @@ namespace FallbackLayer
         pCommandList->SetComputeRootUnorderedAccessView(TriangleCountBufferSlot, triangleCountBuffer);
         pCommandList->SetComputeRootUnorderedAccessView(AABBBufferSlot, aabbBuffer);
         pCommandList->SetComputeRootUnorderedAccessView(InputElementSlot, inputElementBuffer);
-        if (globalDescriptorHeap.ptr)
+        if (baseTreeletsCountBuffer != 0)
         {
-            pCommandList->SetComputeRootDescriptorTable(GlobalDescriptorHeap, globalDescriptorHeap);
+            pCommandList->SetComputeRootUnorderedAccessView(BaseTreeletsCountBufferSlot, baseTreeletsCountBuffer);
+            pCommandList->SetComputeRootUnorderedAccessView(BaseTreeletsIndexBufferSlot, baseTreeletsIndexBuffer);
         }
 
         bool bPrioritizeTrace = buildFlag & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PREFER_FAST_TRACE;
         bool bPrioritizeBuild = buildFlag & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PREFER_FAST_BUILD;
-        UINT numOptimizationPasses = bPrioritizeTrace ? 3 : bPrioritizeBuild  ? 0 : 1;
+
+        UINT numOptimizationPasses;
+
+        if (bPrioritizeBuild)
+        {
+            numOptimizationPasses = 0;
+        }
+        else if (bPrioritizeTrace)
+        {
+            numOptimizationPasses = 3;
+        }
+        else
+        {
+            numOptimizationPasses = 1;
+        }
 
         for (UINT i = 0; i < numOptimizationPasses; i++)
         {
+            if (constants.MinTrianglesPerTreelet > numElements)
+            {
+                break;
+            }
+
             pCommandList->SetComputeRoot32BitConstants(ConstantsSlot, SizeOfInUint32(InputConstants), &constants, 0);
-            const UINT dispatchWidth = DivideAndRoundUp<UINT>(numElements, THREAD_GROUP_1D_WIDTH);
+            
+            auto uavBarrier = CD3DX12_RESOURCE_BARRIER::UAV(nullptr);
+            
+            UINT numGroupsForElements = DivideAndRoundUp<UINT>(numElements, THREAD_GROUP_1D_WIDTH);
+            UINT maxNumTreelets = (UINT) std::max(numElements / constants.MinTrianglesPerTreelet, 1u);
 
             pCommandList->SetPipelineState(m_pClearBufferPSO);
-            pCommandList->Dispatch(dispatchWidth, 1, 1);
-            auto uavBarrier = CD3DX12_RESOURCE_BARRIER::UAV(nullptr);
+            pCommandList->Dispatch(numGroupsForElements, 1, 1);
```

Comment:
Any reason m_pClearBufferPSO and m_pFindTreeletPSO couldn't be combined into a single compute shader?

Reply:
Actually I can't combine them since I need clearBuffers to be fully complete before starting findTreelets